### ISSUE TITLE
Error: One of ['cidr_blocks', ... and Error: [WARN] A duplicate Security Group

### ISSUE THREAD

**Issue by v-stickykeys** [terraform-aws-modules/terraform-aws-security-group#193]
**THREE** things

(1)

Is there some timing issue with creation of security groups? It seems as if it errors, creates it, then tries to create it again.

(2)

If I'm using the nfs module why do I have to specify the nfs rule again?

```
### This is what I __expected__ would work ###

module "nfs_security_group" {
  source  = "terraform-aws-modules/security-group/aws//modules/nfs"
  version = "~> 3.0"

  name        = "${local.namespace}-nfs"
  vpc_id      = var.vpc_id

  source_security_group_id = "${module.grafana_security_group.this_security_group_id}"

  tags = local.default_tags
}
```

(3)

I've looked at the examples already and I still don't fully understand. New to this so please be patient with me 😅 
This is related to #191 which is closed but I think should still be open

```
Error: One of ['cidr_blocks', 'ipv6_cidr_blocks', 'self', 'source_security_group_id', 'prefix_list_ids'] must be set to create an AWS Security Group Rule

  on .terraform/modules/grafana.nfs_security_group/main.tf line 58, in resource "aws_security_group_rule" "ingress_rules":
  58: resource "aws_security_group_rule" "ingress_rules" {



Error: [WARN] A duplicate Security Group rule was found on (sg-0a99bd4d0b4992780). This may be
a side effect of a now-fixed Terraform issue causing two security groups with
identical attributes but different source_security_group_ids to overwrite each
other in the state. See https://github.com/hashicorp/terraform/pull/2376 for more
information and instructions for recovery. Error message: the specified rule "peer: sg-0a99bd4d0b4992780, ALL, ALLOW" already exists

  on .terraform/modules/grafana.nfs_security_group/main.tf line 363, in resource "aws_security_group_rule" "ingress_with_self":
 363: resource "aws_security_group_rule" "ingress_with_self" {
```

```
### This is what produced the above ^ errors ###

module "nfs_security_group" {
  source  = "terraform-aws-modules/security-group/aws//modules/nfs"
  version = "~> 3.0"

  name        = "${local.namespace}-nfs"
  vpc_id      = var.vpc_id

  ingress_with_self = [{
    rule = "all-all"
  }]

  ingress_with_source_security_group_id = [
    {
      rule                     = "nfs-tcp",
      source_security_group_id = "${module.grafana_security_group.this_security_group_id}"
    }
  ]

  tags = local.default_tags
}
```

**Comment by nsvijay04b1** [comment]
I also face the same issue , please advise.  I cannot provide CIDR at all for a RDS DB. 

**module.sg.aws_security_group_rule.ingress_rules[0]: Creating...

Error: One of ['cidr_blocks', 'ipv6_cidr_blocks', 'self', 'source_security_group_id', 'prefix_list_ids'] must be set to create an AWS Security Group Rule**


module "sg" {
  source = "../terraform-aws-security-group-master"
  name                = "${var.identifier}-sg"
  description         = "Security group for RDS"
  vpc_id              = var.vpc_id
  ingress_rules       = ["postgresql-tcp"]
  ingress_with_source_security_group_id =[{
            rule                     = "postgresql-tcp"
            source_security_group_id = sg
            }]
  egress_rules = ["all-tcp"]
  tags         = var.tags
}

**Comment by nsvijay04b1** [comment]
> @antonbabenko
> 
> I also face the same issue , please advise. I cannot provide CIDR at all for a RDS DB.
> 
> **module.sg.aws_security_group_rule.ingress_rules[0]: Creating...**
> **Error: One of ['cidr_blocks', 'ipv6_cidr_blocks', 'self', 'source_security_group_id', 'prefix_list_ids'] must be set to create an AWS Security Group Rule**
> 
> module "sg" {
> source = "../terraform-aws-security-group-master"
> name = "${var.identifier}-sg"
> description = "Security group for RDS"
> vpc_id = var.vpc_id
> ingress_rules = ["postgresql-tcp"]
> ingress_with_source_security_group_id =[{
> rule = "postgresql-tcp"
> source_security_group_id = sg
> }]
> egress_rules = ["all-tcp"]
> tags = var.tags
> }


****UPDATE  - Issue resolved****
This issue resolved after removing ingress_rules  & ingress_cidr_blocks  both , earlier i gave ingress_rules  which required  One of ['cidr_blocks', 'ipv6_cidr_blocks', 'self', 'source_security_group_id', 'prefix_list_ids'] 
  #ingress_cidr_blocks = ["10.0.0.0/24"]
  #ingress_rules       = ["postgresql-tcp"]

**Comment by valmack** [comment]
@nsvijay04b1 Can you paste your working code?

**Comment by henderjm** [comment]
I was testing out this module but kept running into this problem whenever I tried adding another CIDR Block to "ingress_cidr_blocks". I've reverted to just using the Terraform resource.

**Comment by rgarrigue** [comment]
I had a similar issue

```
Error: One of ['cidr_blocks', 'ipv6_cidr_blocks', 'self', 'source_security_group_id', 'prefix_list_ids'] must be set to create an AWS Security Group Rule

  on ../../main.tf line 58, in resource "aws_security_group_rule" "ingress_rules":
  58: resource "aws_security_group_rule" "ingress_rules" {
```

Except I shouldn't go in the mentioned code starting at line 58, I don't have a ingress_rules set, only ingress_with_source_security_group_id.

I worked around this setting a ingress_cidr_blocks = ["my_vpc_cidr"]

**Comment by DanHam** [comment]
## TL;DR

This issue occurs for me when using a ***submodule*** in conjunction with `ingress_with_source_security_group_id`.

Changing the code slightly to use the **main module** rather than a submodule allows Terraform to create the security groups without error.

## Example Code that Causes an Error

The following code causes the error to occur. **Note that the code uses the `http-80` submodule**:

```hcl
module "sg_webserver" {
  source  = "terraform-aws-modules/security-group/aws//modules/http-80"
  version = "4.3.0"

  name        = "webserver-http"
  description = "Allow http requests to webservers from loadbalancer"
  vpc_id      = var.vpc_id

  ingress_with_source_security_group_id = [
    {
      rule                     = "http-80-tcp"
      source_security_group_id = module.sg_loadbalancer.security_group_id
    }
  ]
}
```

```bash
$ terraform apply
#
# Output omitted for brevity
#
╷
│ Error: One of ['cidr_blocks', 'ipv6_cidr_blocks', 'self', 'source_security_group_id', 'prefix_list_ids'] must be set to create an AWS Security Group Rule
│
│   with module.sg_webserver.module.sg.aws_security_group_rule.ingress_rules[0],
│   on .terraform/modules/sg_webserver/main.tf line 54, in resource "aws_security_group_rule" "ingress_rules":
│   54: resource "aws_security_group_rule" "ingress_rules" {
│
╵
```
Note that no error is seen when running `terraform plan`. The error only occurs when executing `terraform apply`. Like @v-stickykeys I felt the above code would work. Put another way, there is nothing in the modules documentation to suggest that this shouldn't work.

## Example Working Code

Changing things around slightly to use the ***main module*** rather than a submodule, allows Terraform to create the security groups **without error**.

**The following code works as expected.** Again, note that the code uses the main module - **not a submodule**.

```hcl
module "sg_webserver" {
  source  = "terraform-aws-modules/security-group/aws"
  version = "4.3.0"

  name        = "webserver-http"
  description = "Allow http requests to webservers from loadbalancer"
  vpc_id      = var.vpc_id

  ingress_with_source_security_group_id = [
    {
      rule                     = "http-80-tcp"
      source_security_group_id = module.sg_loadbalancer.security_group_id
    }
  ]

  egress_rules = ["all-all"]
}
```

The inclusion of `egress_rules` causes the code above to work the same way as the submodule (the http-80 submodule creates egress rules automatically; the main module does not).

**Comment by khdevel** [comment]
Thank you @DanHam for the hint with _do not use the submodule_. I am using `ingress_with_source_security_group_id` as well as `egress_with_source_security_group_id` and I spent a lot of time investigating the problem! After you suggestion everything is working now as expected!

**Comment by calebtote** [comment]
This caused a lot of headache for me and unfortunately makes the submodules directly unusable, which is a shame.
